libraries/dl-streamer/src/monolithic/gst/elements/gvamotiondetect/gvamotiondetect.cpp

```diff
@@ -0,0 +1,758 @@
+
+    cv::UMat thresh_small;
+    const int PIXEL_DIFF_THR = 15; // same threshold (on 0..255 grayscale)
+    cv::threshold(blurred_small, thresh_small, PIXEL_DIFF_THR, 255, cv::THRESH_BINARY);
+
+    // Light morphology to reduce noise
+    cv::UMat morph_small;
+    cv::Mat ksmall = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(3, 3));
+    cv::morphologyEx(thresh_small, morph_small, cv::MORPH_OPEN, ksmall);
+    cv::dilate(morph_small, morph_small, ksmall, cv::Point(-1, -1), 1);
+
+    // Block-based scan instead of contours
+    std::vector<MotionRect> rois;
+    const double MIN_REL_AREA = 0.0005; // fraction of full image (tunable)
+    double full_area = (double)width * (double)height;
+    double scale_x = (double)width / (double)small_w;
+    double scale_y = (double)height / (double)small_h;
+
+    int block_full = std::max(16, self->block_size); // clamp
+    int block_small_w = std::max(4, (int)std::round(block_full / scale_x));
+    int block_small_h = std::max(4, (int)std::round(block_full / scale_y));
+    const double CHANGE_RATIO_THR =
+        std::max(0.0, std::min(1.0, self->motion_threshold)); // fraction of changed pixels within block (property)
+    cv::Mat morph_cpu = morph_small.getMat(cv::ACCESS_READ);
+
+    for (int by = 0; by < small_h; by += block_small_h) {
+        int h_small = std::min(block_small_h, small_h - by);
+        if (h_small < 4)
+            break;
+        for (int bx = 0; bx < small_w; bx += block_small_w) {
+            int w_small = std::min(block_small_w, small_w - bx);
+            if (w_small < 4)
+                break;
+            cv::Rect r_small(bx, by, w_small, h_small);
+            cv::Mat sub = morph_cpu(r_small);
+            int changed = cv::countNonZero(sub);
+            double ratio = (double)changed / (double)(r_small.width * r_small.height);
+            if (ratio < CHANGE_RATIO_THR)
+                continue; // insufficient motion in block
+
+            // Scale to full-res
+            int fx = (int)std::round(r_small.x * scale_x);
+            int fy = (int)std::round(r_small.y * scale_y);
+            int fw = (int)std::round(r_small.width * scale_x);
+            int fh = (int)std::round(r_small.height * scale_y);
+
+            double area_full = (double)fw * (double)fh;
+            if (area_full / full_area < MIN_REL_AREA)
+                continue; // too small
+
+            // Pad a little
+            const int PAD = 4;
+            fx = std::max(0, fx - PAD);
+            fy = std::max(0, fy - PAD);
+            fw = std::min(width - fx, fw + 2 * PAD);
+            fh = std::min(height - fy, fh + 2 * PAD);
+
+            // Clamp
+            if (fx + fw > width)
+                fw = width - fx;
+            if (fy + fh > height)
+                fh = height - fy;
+
+            rois.push_back(MotionRect{fx, fy, fw, fh});
+        }
+    }
+
+    // Merge overlapping ROIs (simple O(n^2) since counts expected low)
```

Comment:
The outer while-loop makes it O(N^3), isn't it?
You could use "Disjoint Set Union (DSU) / Union-Find" instead, which is O(N^2).
Or just use "cv::groupRectangles()" (requiring to use cv::Rect instead of MotionRect)?